QualifiedTeachersApi/src/QualifiedTeachersApi/Infrastructure/OpenApi/ServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,121 @@
+{
+    public static IServiceCollection AddOpenApi(this IServiceCollection services, IConfiguration configuration)
+    {
+        services.AddExampleProviders(typeof(Program).Assembly);
+
+        foreach (var version in Api.Constants.Versions)
+        {
+            services.AddOpenApiDocument((settings, provider) =>
+            {
+                settings.DocumentName = $"Qualified Teachers API {version}";
+                settings.Version = version;
+                settings.Title = "Qualified Teachers API";
+                settings.ApiGroupNames = new[] { version };
+                settings.AddExamples(provider);
+                settings.TypeNameGenerator = new GeneratedCrmTypeNameGenerator(settings.TypeNameGenerator);
+
+                settings.DocumentProcessors.Add(new PopulateResponseDescriptionOperationProcessor());
+
+                settings.SchemaProcessors.Add(new RemoveEnumValuesSchemaProcessor());
+
+                settings.OperationProcessors.Add(new ResponseContentTypeOperationProcessor());
+                settings.OperationProcessors.Add(new AddTooManyRequestsResponseOperationProcessor());
+                settings.OperationProcessors.Add(new PopulateResponseDescriptionOperationProcessor());
+                settings.OperationProcessors.Add(new AssignBinaryContentTypeFromProducesOperationProcessor());
+
+                settings.SchemaGenerator = new HandleOptionJsonSchemaGenerator(settings);
+
+                settings.AddSecurity(SecuritySchemes.ApiKey, new NSwag.OpenApiSecurityScheme()
+                {
+                    In = NSwag.OpenApiSecurityApiKeyLocation.Header,
+                    Name = "Authorization",
+                    Scheme = "Bearer",
+                    Type = NSwag.OpenApiSecuritySchemeType.Http
+                });
+
+                if (version == "v3")
+                {
+                    // Only V3 uses ID access tokens for auth
+                    settings.AddSecurity(SecuritySchemes.GetAnIdentityAccessToken, new NSwag.OpenApiSecurityScheme()
+                    {
+                        In = NSwag.OpenApiSecurityApiKeyLocation.Header,
+                        Scheme = "Bearer",
+                        Type = NSwag.OpenApiSecuritySchemeType.OpenIdConnect,
+                        OpenIdConnectUrl = configuration["GetAnIdentity:BaseAddress"] + ".well-known/openid-configuration"
+                    });
+
+                    // Add operation-level security scheme instead of document-level
+                    settings.OperationProcessors.Add(new AddSecuritySchemeOperationProcessor());
+                }
+                else
+                {
+                    settings.DocumentProcessors.Add(new AddApiKeySecuritySchemeDocumentProcessor());
+                }
+            });
+        }
+
+        services.AddSingleton<IStartupFilter, OpenApiEndpointsStartupFilter>();
+
+        return services;
+    }
+}
+
+public class OpenApiEndpointsStartupFilter : IStartupFilter
```

Comment:
Not in a separate file ?

Reply:
There's a lot of cross-over between the two types so I felt having them in a single file would be easier.

Reply:
There's a follow-up PR coming where there's a lot less cross-over; I'll look to move it out into its own file with that change.